crates/bevy_ecs/src/system/query.rs

```diff
@@ -11,102 +11,216 @@ use bevy_tasks::TaskPool;
 use std::{any::TypeId, fmt::Debug};
 use thiserror::Error;
 
-/// Provides scoped access to a [`World`] according to a given [`WorldQuery`] and query filter.
+/// Provides scoped access to components in a [`World`].
 ///
-/// Queries are a powerful tool enabling the programmer to iterate over entities and their components
-/// as well as filtering them on certain conditions.
+/// Queries allow to iterate over entities and their components as well as filtering them
+/// on certain conditions. A query matches its parameters against the world to produce a series
+/// of results. Each *query result* is a tuple of components (the same components defined
+/// in the query) that belong to the same entity.
 ///
```

Comment:
I think its worth calling out that Queries have an internal archetype cache to avoid re-computing archetype matches on each query access. This will help people reason about the "cost" of queries.